### PR TITLE
Script to create Smile-Warrior dataset, script to load data correctly and share it and script essential for data augmentation

### PR DESCRIPTION
**Pull request by Kecz**
Updating README and adding following scripts:

adjust_dataset.py - this script creates dataset for Smile-Warrior from fer2013.csv dataset for for Facial Expression Recognition.

data_server.py - this script contains 2 functions. Funtion load_dataset i used to process loaded dataset and divide it into train, test and validate parts. Funtion show_picture allows to show selected image from dataset.

data_augmentation.py - script that creates ImageDataGenerator used in process of training.

### CODE REVIEW

File: adjust_dataset.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+#Script changing labels of emotions to 1 - smile, 0 - no smile and saving changed dataset to new csv file
+
+
+def main():
+    import pandas as pd
+
+    HAPPINESS = 3
+
+    data_set = pd.read_csv("fer2013.csv")
```
Comment (Kecz) [inline adjust_dataset.py]: Later I will be working on how to pass path to csv from command line argument.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,14 @@
 # smile-warrior
 This repo contains code used for smile detector for project of Gradient Science Club.
+
+To create dataset run: \
+    adjust_dataset.py
```
Comment (pawelkopec) [inline README.md]: Rather:
```
python adjust_dataset.py
```
Same below.

File: README.md [hunk 3]
```diff
@@ -1,2 +1,14 @@
...
+    adjust_dataset.py 
+
+Result will be csv file named 'smile_warrior_dataset'.
+
+To load dataset use function:\
+    load_dataset  
+from module:\
+    data_server.py
+    
+Which returns respectively: X_train, Y_train, X_test, Y_test, X_validate and Y_validate datasets.
```
Comment (pawelkopec) [inline README.md]: User of this repo should not be forced to use any functions, just already working scripts.

File: adjust_dataset.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+# Script changing labels of emotions to 1 - smile, 0 - no smile and saving changed dataset to new csv file
+import pandas as pd
+import argparse
+
+HAPPINESS = 3
+
+
+def main():
+
+    #data_set = pd.read_csv("fer2013.csv")
```
Comment (pawelkopec) [inline adjust_dataset.py]: You can remove it.

File: adjust_dataset.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
+# Script changing labels of emotions to 1 - smile, 0 - no smile and saving changed dataset to new csv file
+import pandas as pd
+import argparse
+
+HAPPINESS = 3
+
+
+def main():
```
Comment (pawelkopec) [inline adjust_dataset.py:7]: This is minor, but please separate functions:
```
def adjust_dataset(input_file):
    ...

def parse_args():
    ...
    return parser.parse_args()

File: data_server.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+        for line in file:
+
+            row = line.split(',')
+
+            # Pixel values for each picture are saved in csv in single cells as a string containing values splited
+            # by space-bar so we need to split that string to substrings, convert them to ints and add these pixel
+            # values to list one by one
+
+            if row[2] == "Training\n":
```
Comment (pawelkopec) [inline data_server.py]: Nit - you can put thos strings into const for clarity.

File: data_server.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        Y_train = np.array(Y_list_train)
+        Y_test = np.array(Y_list_test)
+        Y_validate = np.array(Y_list_valid)
+
+        print("Dataset prepared")
+
+        return X_train, Y_train, X_test, Y_test, X_validate, Y_validate
+
+
+# Showing single picture from X_data numpy array, number of picture is controled by variable 'which_one'
```
Comment (pawelkopec) [inline data_server.py]: Please add comments for a function as pydocstring:
https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

File: data_server.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+def show_picture(which_one, x_data):
```
Comment (pawelkopec) [inline data_server.py]: Why do I need to pass x_data to show only one picture. A picture should be an argument here. Index can be infered outside of this function. What if I have one picture only and I want to use it?

File: data_server.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+    plt.imshow(x_data[which_one, :].reshape(48, 48))
```
Comment (pawelkopec) [inline data_server.py]: 48 hardcoded. You should infer the size or pass it as an argument.

File: example.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
+# Example how to load dataset by function from module and to show single picture
+
+import sys
+import argparse
+
+# You may need to execute line below:
+# sys.path.append("Your path to project file for example: B:/Gradient/Smile_warrior")
```
Comment (pawelkopec) [inline example.py]: No, this should be handled by PYTHONPATH and not present in the code.

File: example.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
+# Example how to load dataset by function from module and to show single picture
```
Comment (pawelkopec) [inline example.py]: Not sure if this example is needed. Description of how to use it should by rather in pydocstring of load_dataset(...) function. If you want to have sth to check if it works you can use pytest and create a test from it.

File: README.md [hunk 12]
```diff
@@ -1,2 +1,21 @@
 # smile-warrior
 This repo contains code used for smile detector for project of Gradient Science Club.
+
```
Comment (Kecz) [inline README.md:3]: test comment

File: README.md (hunk 2 above)
Comment (Kecz) [reply README.md]: Ok

File: adjust_dataset.py (hunk 5 above)
Comment (Kecz) [reply adjust_dataset.py:7]: Ok

File: data_server.py (hunk 6 above)
Comment (Kecz) [reply data_server.py]: Do you mean creating for example global list: 
Usage = ["Training\n", "PublicTest\n", "PrivateTest\n"] 
and then referring to that list in code like that?: 
if row[2] == Usage[0]

File: README.md (hunk 3 above)
Comment (Kecz) [reply README.md]: Could you explain what you mean more precisely, please? Should the user only import data_server.py instead of calling load_dataset() from data_server.py?

File: example.py (hunk 10 above)
Comment (Kecz) [reply example.py]: Ok

File: example.py (hunk 11 above)
Comment (Kecz) [reply example.py]: Ok, I will move it to pydocstring

File: data_server.py (hunk 9 above)
Comment (Kecz) [reply data_server.py]: Ok

File: data_server.py (hunk 8 above)
Comment (Kecz) [reply data_server.py]: Ok

File: data_server.py (hunk 7 above)
Comment (Kecz) [reply data_server.py]: I have done it, please check if these pydocstrings are missing some elements or are too detailed.

File: adjust_dataset.py [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    To run correctly you need to pass two paths to csv files as a command-line arguments, first for input file which
+    will be adjusted for Smile-Warrior needs and second for output file where new dataset will be created.
+
+    Example:
+
+        $ python adjust_dataset.py fer2013.csv smile_warrior_dataset.csv
+
+    """
+
+    if path_from_terminal == 0:
```
Comment (pawelkopec) [inline adjust_dataset.py]: This looks bad to give a default path to a file. Let's keep it simple. In README we have it as an argument anyway.

File: adjust_dataset.py [hunk 14]
```diff
@@ -0,0 +1,86 @@
+import pandas as pd
+import argparse
+
+HAPPINESS = 3
+
+# Alternative for users who don't pass arguments from command-line, that alternative will be deleted in future version
+path_input_file = "fer2013.csv"
+path_output_file = 'smile_warrior_dataset.csv'
+
+# Set that flag to 1 if you want to pass paths for csv files as a command-line arguments
```
Comment (pawelkopec) [inline adjust_dataset.py]: This looks like an overkill, cli argument will suffice. Keep It Simple Stupid. And for future - use booleans instead of 0/1.

File: adjust_dataset.py [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+    # Emotion with label HAPPINESS is happiness so we change label to 1 for our smile detector, other emotions are
+    # changed to 0.
+    for index, row in data_set.iterrows():
+        if row['emotion'] == HAPPINESS:
+            data_set.at[index, 'emotion'] = 1
+        else:
+            data_set.at[index, 'emotion'] = 0
+
+    if path_from_terminal == 0:
```
Comment (pawelkopec) [inline adjust_dataset.py]: Same a above.

File: adjust_dataset.py [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+    Function to get paths to csv files from command-line arguments
+
+    Example:
+
+        args = parse_args()
+
+    :return:
+        Parser with two arguments 'input' and 'output'.
+
+        First argument 'input' is path to csv file with dataset which will be adjusted for Smile-Warrior needs.
```
Comment (pawelkopec) [inline adjust_dataset.py]: Leave it. Those are not the arguments of this function.

File: adjust_dataset.py [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+        print("Smile-Warrior dataset created in file %s" % args.output)
+
+
+def parse_args():
+    """
+    Function to get paths to csv files from command-line arguments
+
+    Example:
+
+        args = parse_args()
```
Comment (pawelkopec) [inline adjust_dataset.py]: This is not needed at all. Use examples when there is more to do than an empty call ().

File: README.md [hunk 18]
```diff
@@ -1,2 +1,21 @@
...
+```
+   python adjust_dataset.py 
+```
+
+Result will be csv file named 'smile_warrior_dataset'.
+
+## To load dataset:
+Use function:
+```
+    load_dataset
```
Comment (pawelkopec) [inline README.md]: As I said already - do not mention functions you created in README. It should be transparent to the user which functions we use in training. If user wants to use the function, he will go to see the commentary within that file.

File: data_augmentation.py [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+
+
+def create_datagen():
+    """
+    Function creates ImageDataGenerator used in process of augmentation Smile-Warrior dataset to increase accuracy of
+    trained neural network. Arguments of ImageDataGenerator can be adjusted below.
+
+    Example:
+
+            datagen = create_datagen()
```
Comment (pawelkopec) [inline data_augmentation.py]: This is not useful.

File: data_augmentation.py [hunk 20]
```diff
@@ -0,0 +1,28 @@
+from keras.preprocessing.image import ImageDataGenerator
+
+shift = 0.1
+
+
+def create_datagen():
+    """
+    Function creates ImageDataGenerator used in process of augmentation Smile-Warrior dataset to increase accuracy of
```
Comment (pawelkopec) [inline data_augmentation.py]: I do not think comment is useful here. If user wants to know the details, he will check the docs for ImageDataGenerator.

File: data_server.py [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+import matplotlib.pyplot as plt
+import math
+
+usage = ["Training\n", "PublicTest\n", "PrivateTest\n"]
+
+
+def load_dataset(filename):
+    """
+    Function loads dataset from csv file at path passed as param 'filename' and then returns respectively:
+    x_train, y_train, x_test, y_test, x_validate and y_validate datasets.
```
Comment (pawelkopec) [inline data_server.py]: This is already mentioned twice in just this comment.

File: data_server.py [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+
+def show_picture(picture, x_dim, y_dim):
+    """
+    Showing single picture passed as an input param in form of a numpy array.
+
+    Function reshapes input numpy array to shape (x_dim, y_dim) and then uses matplotlib.pyplot.imshow to plot it.
+
+    Example:
+
+        show_picture(x_train[9], 48, 48)
```
Comment (pawelkopec) [inline data_server.py]: I can't see how this example helps anyhow.

File: data_server.py [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+        Dimension of showed picture in X axis.
+
+    :param y_dim:
+        Dimension of showed picture in Y axis.
+    """
+
+    plt.imshow(picture.reshape(x_dim, y_dim))
+    plt.show()
+
+
```
Comment (pawelkopec) [inline data_server.py]: Extra newline.

File: README.md (hunk 18 above)
Comment (Kecz) [reply README.md]: Ok

File: adjust_dataset.py (hunk 14 above)
Comment (Kecz) [reply adjust_dataset.py]: Ok

File: adjust_dataset.py (hunk 13 above)
Comment (Kecz) [reply adjust_dataset.py]: Ok

File: adjust_dataset.py (hunk 17 above)
Comment (Kecz) [reply adjust_dataset.py]: Ok

File: adjust_dataset.py (hunk 16 above)
Comment (Kecz) [reply adjust_dataset.py]: OK

File: data_augmentation.py (hunk 19 above)
Comment (Kecz) [reply data_augmentation.py]: Ok

File: README.md [hunk 24]
```diff
@@ -1,2 +1,17 @@
...
+```
+
+Result will be a csv file with dataset for Smile-Warrior (by default in: "smile_warrior_dataset.csv").
+
+## To load dataset:
+Use:
+```
+   python data_server.py
+```
+Which returns respectively: x_train, y_train, x_test, y_test, x_validate and y_validate datasets.
```
Comment (pawelkopec) [inline README.md]: What do you mean that script returns datasets?

File: adjust_dataset.py [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+    Expression Recognition.
+
+    This script is changing labels of emotions in fer2013.csv to 1 - smile, 0 - no smile.
+
+    To run correctly you need to pass two paths to csv files as a command-line arguments, first for input file which
+    will be adjusted for Smile-Warrior needs and second for output file where new dataset will be created.
+
+    Example:
+
+        $ python adjust_dataset.py -input fer2013.csv -output smile_warrior_dataset.csv
```
Comment (pawelkopec) [inline adjust_dataset.py]: Not sure if needed. Argparser already gives help on how to use it.

Comment (Kecz) [reply adjust_dataset.py]: Ok, I removed it.

File: README.md (hunk 24 above)
Comment (Kecz) [reply README.md]: Do you mean that I should change it to something like "Which provides us with function that returns respectively: x_train, y_train, x_test, y_test, x_validate and y_validate datasets."?

Comment (pawelkopec) [reply README.md]: User does not need to know the function within script. He just runs:
```
python data_server.py
```
Why mention function at all?

File: data_augmentation.py [hunk 26]
```diff
@@ -0,0 +1,17 @@
+from keras.preprocessing.image import ImageDataGenerator
+
+shift = 0.1
```
Comment (pawelkopec) [inline data_augmentation.py]: Why global?

File: adjust_dataset.py [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+    data_set.to_csv(args.output, index=False)
+    print("Smile-Warrior dataset created in file %s" % args.output)
+
+
+def parse_args():
+    """
+    Function to get paths to csv files from command-line arguments
+
+    :return:
+        Parser with two arguments 'input' and 'output'.
```
Comment (pawelkopec) [inline adjust_dataset.py]: Why mention this?

**Review by pawelkopec — APPROVED** [review]

File: adjust_dataset.py (hunk 27 above)
Comment (Kecz) [reply adjust_dataset.py]: Ok, I removed it.

File: data_augmentation.py (hunk 26 above)
Comment (Kecz) [reply data_augmentation.py]: Changed it to local.

File: README.md (hunk 24 above)
Comment (Kecz) [reply README.md]: Ok, I removed info about function and rewrote it more generally
